File: src/VBNetFx/FacessoData/SQLServer/SPAccess_BonusList.vb
Imports Activedev
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.Common

Partial Public NotInheritable Class SPAccess

    Public Sub BonusList_AddEntry(ByVal bli As BonusListItem, ByVal IDUserCreated As Integer)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("BonusList_AddEntry", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = bli.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = bli.CostCenterInfo.IDCostCenter
                .Add("@DegreeOfTime", SqlDbType.Float).Value = bli.DegreeOfTime
                .Add("@Percentage", SqlDbType.Float).Value = bli.Percentage
                .Add("@AbsoluteValue", SqlDbType.Float).Value = bli.AbsoluteValue
                .Add("@IDUserCreated", SqlDbType.Int).Value = IDUserCreated
            End With
            locCmd.ExecuteReader()
        End Using
    End Sub

    Public Sub BonusList_CreateBaseList(ByVal IDSubsidiary As Integer, _
                                       ByVal IDCostCenter As Integer, _
                                       ByVal IDUserCreated As Integer)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("BonusList_CreateBaseList", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = IDCostCenter
                .Add("@IDUserCreated", SqlDbType.Int).Value = IDUserCreated
            End With
            locCmd.ExecuteReader()
        End Using
    End Sub

    Public Sub BonusList_DeleteList(ByVal IDSubsidiary As Integer, _
                                       ByVal IDCostCenter As Integer, _
                                       ByVal IDUserCalled As Integer)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("BonusList_DeleteList", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = IDCostCenter
                .Add("@IDUserCalled", SqlDbType.Int).Value = IDUserCalled
            End With
            locCmd.ExecuteReader()
        End Using
    End Sub

    Public Sub BonusList_FromBaseCostCenter(ByVal IDSubsidiary As Integer, _
                                               ByVal IDBaseCostCenter As Integer, _
                                               ByVal ForIDCostCenter As Integer, _
                                               ByVal IDUserCalled As Integer)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("BonusList_FromBaseCostCenter", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = IDSubsidiary
                .Add("@IDBaseCostCenter", SqlDbType.Int).Value = IDBaseCostCenter
                .Add("@IDCostCenter", SqlDbType.Int).Value = ForIDCostCenter
                .Add("@IDUserCalled", SqlDbType.Int).Value = IDUserCalled
            End With
            locCmd.ExecuteReader()
        End Using
    End Sub

    Public Sub BonusList_ReplaceEntry(ByVal bli As BonusListItem, ByVal IDUserCalled As Integer)

        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return
        End If
        Using locConnection
            Dim locCmd As New SqlCommand("BonusList_ReplaceEntry", locConnection)
            locCmd.CommandType = CommandType.StoredProcedure

            With locCmd.Parameters
                .Add("@IDSubsidiary", SqlDbType.Int).Value = bli.IDSubsidiary
                .Add("@IDCostCenter", SqlDbType.Int).Value = bli.CostCenterInfo.IDCostCenter
                .Add("@DegreeOfTime", SqlDbType.Float).Value = bli.DegreeOfTime
                .Add("@Percentage", SqlDbType.Float).Value = bli.Percentage
                .Add("@AbsoluteValue", SqlDbType.Float).Value = bli.AbsoluteValue
                .Add("@IDUserCalled", SqlDbType.Int).Value = IDUserCalled
            End With
            locCmd.ExecuteReader()
        End Using
    End Sub

    Public Function BonusList_GetCostCenterInfoCollection(ByVal IDSubsidiary As Integer, ByVal Invert As Boolean) As CostcenterInfoItems
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommandString As String = "SELECT DISTINCT CostCenters.*" & _
                    "FROM CostCenters INNER JOIN " & _
                     "BonusLists ON BonusLists.IDSubsidiary = CostCenters.IDSubsidiary "
            If Invert Then
                locCommandString &= "AND BonusLists.IDCostCenter <> CostCenters.IDCostCenter WHERE CostCenters.IDSubsidiary=" & IDSubsidiary
            Else
                locCommandString &= "AND BonusLists.IDCostCenter = CostCenters.IDCostCenter WHERE CostCenters.IDSubsidiary=" & IDSubsidiary
            End If

            Dim locCommand As New SqlCommand(locCommandString, locConnection)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                Dim locUIC As New CostcenterInfoItems
                Do While locDR.Read
                    Dim locCostcenterInfo As New CostcenterInfo(locDR)
                    locUIC.Add(locCostcenterInfo)
                Loop
                Return locUIC
            Else
                Return Nothing
            End If
        End Using
    End Function

    Public Function BonusList_GetBonusListItems(ByVal IDSubsidiary As Integer, ByVal IDCostCenter As Integer) As BonusListItems
        Dim locConnection As SqlConnection = GetOpenedConnectionSafely()
        If locConnection Is Nothing Then
            Return Nothing
        End If
        Using locConnection
            Dim locCommandString As String = "SELECT [BonusList].* FROM [BonusList] INNER JOIN [BonusLists] " & _
                      "ON BonusList.IDBonusLists = BonusLists.IDBonusLists AND BonusList.IDSubsidiary = BonusLists.IDSubsidiary " & _
                      " WHERE [BonusLists].[IDSubsidiary]=" & IDSubsidiary & " AND [BonusLists].[IDCostCenter]=" & IDCostCenter & _
                      " ORDER BY [DegreeOfTime]"

            Dim locCommand As New SqlCommand(locCommandString, locConnection)
            Dim locCci As CostcenterInfo = SPAccess.GetInstance.GetCostCenter(IDSubsidiary, IDCostCenter)
            Dim locDR As SqlDataReader = locCommand.ExecuteReader()
            If locDR.HasRows Then
                Dim locBlis As New BonusListItems
                Do While locDR.Read
                    Dim locBli As New BonusListItem(locDR, locCci)
                    locBlis.Add(locBli)
                Loop
                Return locBlis
            Else
                Return Nothing
            End If
        End Using
    End Function
End Class

File: src/MixNetFx/FacessoData/SQLServer/SPAccess_BonusList.cs
using System;
using System.Data;
using System.Data.SqlClient;
using ActiveDev;

namespace Facesso.Data
{
    public sealed partial class SPAccess
    {
        public void BonusList_AddEntry(BonusListItem bli, int idUserCreated)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("BonusList_AddEntry", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = bli.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = bli.CostCenterInfo.IDCostCenter;
                locCmd.Parameters.Add("@DegreeOfTime", SqlDbType.Float).Value = bli.DegreeOfTime;
                locCmd.Parameters.Add("@Percentage", SqlDbType.Float).Value = bli.Percentage;
                locCmd.Parameters.Add("@AbsoluteValue", SqlDbType.Float).Value = bli.AbsoluteValue;
                locCmd.Parameters.Add("@IDUserCreated", SqlDbType.Int).Value = idUserCreated;
                locCmd.ExecuteReader();
            }
        }

        public void BonusList_CreateBaseList(int idSubsidiary, int idCostCenter, int idUserCreated)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("BonusList_CreateBaseList", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = idCostCenter;
                locCmd.Parameters.Add("@IDUserCreated", SqlDbType.Int).Value = idUserCreated;
                locCmd.ExecuteReader();
            }
        }

        public void BonusList_DeleteList(int idSubsidiary, int idCostCenter, int idUserCalled)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("BonusList_DeleteList", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = idCostCenter;
                locCmd.Parameters.Add("@IDUserCalled", SqlDbType.Int).Value = idUserCalled;
                locCmd.ExecuteReader();
            }
        }

        public void BonusList_FromBaseCostCenter(int idSubsidiary, int idBaseCostCenter, int forIDCostCenter, int idUserCalled)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("BonusList_FromBaseCostCenter", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = idSubsidiary;
                locCmd.Parameters.Add("@IDBaseCostCenter", SqlDbType.Int).Value = idBaseCostCenter;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = forIDCostCenter;
                locCmd.Parameters.Add("@IDUserCalled", SqlDbType.Int).Value = idUserCalled;
                locCmd.ExecuteReader();
            }
        }

        public void BonusList_ReplaceEntry(BonusListItem bli, int idUserCalled)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return;
            using (locConnection)
            {
                var locCmd = new SqlCommand("BonusList_ReplaceEntry", locConnection);
                locCmd.CommandType = CommandType.StoredProcedure;
                locCmd.Parameters.Add("@IDSubsidiary", SqlDbType.Int).Value = bli.IDSubsidiary;
                locCmd.Parameters.Add("@IDCostCenter", SqlDbType.Int).Value = bli.CostCenterInfo.IDCostCenter;
                locCmd.Parameters.Add("@DegreeOfTime", SqlDbType.Float).Value = bli.DegreeOfTime;
                locCmd.Parameters.Add("@Percentage", SqlDbType.Float).Value = bli.Percentage;
                locCmd.Parameters.Add("@AbsoluteValue", SqlDbType.Float).Value = bli.AbsoluteValue;
                locCmd.Parameters.Add("@IDUserCalled", SqlDbType.Int).Value = idUserCalled;
                locCmd.ExecuteReader();
            }
        }

        public CostcenterInfoItems BonusList_GetCostCenterInfoCollection(int idSubsidiary, bool invert)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                string locCommandString = "SELECT DISTINCT CostCenters.*" +
                    "FROM CostCenters INNER JOIN " +
                    "BonusLists ON BonusLists.IDSubsidiary = CostCenters.IDSubsidiary ";
                if (invert)
                    locCommandString += "AND BonusLists.IDCostCenter <> CostCenters.IDCostCenter WHERE CostCenters.IDSubsidiary=" + idSubsidiary;
                else
                    locCommandString += "AND BonusLists.IDCostCenter = CostCenters.IDCostCenter WHERE CostCenters.IDSubsidiary=" + idSubsidiary;

                var locCommand = new SqlCommand(locCommandString, locConnection);
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    var locUIC = new CostcenterInfoItems();
                    while (locDR.Read())
                    {
                        var locCostcenterInfo = new CostcenterInfo(locDR);
                        locUIC.Add(locCostcenterInfo);
                    }
                    return locUIC;
                }
                return null;
            }
        }

        public BonusListItems BonusList_GetBonusListItems(int idSubsidiary, int idCostCenter)
        {
            SqlConnection locConnection = GetOpenedConnectionSafely();
            if (locConnection == null) return null;
            using (locConnection)
            {
                string locCommandString = "SELECT [BonusList].* FROM [BonusList] INNER JOIN [BonusLists] " +
                    "ON BonusList.IDBonusLists = BonusLists.IDBonusLists AND BonusList.IDSubsidiary = BonusLists.IDSubsidiary " +
                    " WHERE [BonusLists].[IDSubsidiary]=" + idSubsidiary + " AND [BonusLists].[IDCostCenter]=" + idCostCenter +
                    " ORDER BY [DegreeOfTime]";

                var locCommand = new SqlCommand(locCommandString, locConnection);
                CostcenterInfo locCci = SPAccess.GetInstance().GetCostCenter(idSubsidiary, idCostCenter);
                SqlDataReader locDR = locCommand.ExecuteReader();
                if (locDR.HasRows)
                {
                    var locBlis = new BonusListItems();
                    while (locDR.Read())
                    {
                        var locBli = new BonusListItem(locDR, locCci);
                        locBlis.Add(locBli);
                    }
                    return locBlis;
                }
                return null;
            }
        }
    }
}
